### ISSUE TITLE
platforms: centralise host/platform selection logic

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-flow#5404]
The logic for managing `bad_hosts` is spread throughout the code. This should be centralised to allow for easier configuration of retries for activities such as job polling which is currently difficult and messy.

#### IHS

If a host is not contactable (i.e. SSH / 255 error) we retry on the next host for operations like remote-init. This is not universally supported throughout the Cylc codebase but should be.

#### Bad Hosts

To avoid running operations on hosts we know are down we maintain a set of "bad hosts". Hosts are added to the set when we recieve a 255 error code. The set is drained routinely be a main loop plugin. In the case of job submission, if all hosts are in the "bad hosts" set, we reset all "bad hosts" for the job's platform, in order to allow submission to retry (not included in the platforms proposal document).

So the "bad hosts" logic is effectively this:

* Go through "good hosts" for each supported operation.
* If the operation is job-submission and we run out of hosts, reset all "bad hosts" for the platform.
* If we run out of hosts raise an error.
* On a scheduled interval reset "bad hosts".

#### Implementation Woes

Part of the problem of implementing this logic is that the subprocpool uses a call/callback pattern which unfortunately makes writing retry logic rather difficult.

TLDR; currently retry logic should look something like this:

```python
def do_something():
    self.subprocpool.run(
       cmd,
       do_something_success_callback,
       do_something_255_callback,
       callback_args=state,
    )

def do_something_success_callback(..., state):
    ...

def do_something_255_callback(..., state):
  try:
    # retry
    do_something(...)
  except NoHostsError:
    if state.has_reset_hosts:
      # fail
      raise
    else:
      # reset bad hosts and retry
      self.bad_hosts -= platform.hosts
      do_something(...)
```

This is a handful, getting this implemented for each remote procedure results in a lot of duplication and awkward logic management.

#### Bad Hosts

This modify the bad hosts logic to this:

* For each operation:
  * Go through the "good hosts" first.
  * If we run out of them, go through the "bad hosts".
  * If a bad host is found to be good, remove it from "bad hosts".
  * If we run out of hosts, raise an error.

> **Note:** that using `async` makes it easier to maintain the state of the retry loop making it easier to manage state on a per-operation basis.

> **Note:** The "bad hosts" list is global in both examples so parallel operations can feedback host info to each other.

#### Outline

To achieve this the host/platform selection should be performed in a for loop like this:

```python
def iter_hosts(platform_name):
     # try the good hostsskipped_hosts = []
    for platform in select_platform(platform_group):
        for host in select_host(platform):
            if host in bad_hosts:
                 skipped_hosts.append(host)
            else:
                yield host

    # try the bad hosts
    yield from skipped_hosts

    # fail the operation
    raise NoHostsError(...)
```

Which can be used like this:

```python
async def poll_tasks(itasks):
    for platform_name, tasks in get_platforms(itasks):
        for host in iter_hosts(platform_name):
            with suppress(255Error):
                await _poll_tasks(itasks, host)
        else:
            # all hosts down, handle error
            LOG.error(f'Could not poll tasks on platform {platform}, all hosts unavailable')
```

#### Implementation

This is techincally possible now, however, it's a big change and the call/callback pattern does rather make a mess of things.

Suggest attempting this with #5017 or at least a cut-down version of this which sets the result on future objects in order to allow operations to be written with the `async` paradigm e.g:

```python
class SubprocPool:

    def run(...):
        future = asyncio.Future()
        self._run(...)
        return future

    def _process(...):
        for proc in self._running:
            ...
            if proc.returncode:
                if proc.callback:
                    proc.callback(*proc.callback_args)
                future.set(proc.returncode)
```

#### Task Management

One bit of code which would greatly benefit from this is the job-submission pipeline which currently works by sending the same tasks back through the same method over and over again until they finally fall out the bottom.

```
async def submit_tasks(self, itasks):
    for platform_name, itasks in get_platform(itasks):
        platform_name = await subproc_platform_eval(platform_name)
        await remote_init(platform)
        await remote_file_install(platform)
        await job_submit(platform, itasks)
```
